src/lib.rs

```diff
@@ -1264,6 +1278,85 @@ where
     }
 }
 
+/// A streaming iterator that flattens nested streaming iterators.
+#[derive(Debug)]
+pub struct Flatten<I> {
+    iter: I,
+    first: bool,
+    inner_first: bool,
+}
+
+impl<I, J> StreamingIterator for Flatten<I>
+where
+    I: StreamingIteratorMut<Item = J>,
+    for<'a> J: StreamingIterator + 'a,
+{
+    type Item = J::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        if self.first {
+            self.first = false;
+            self.iter.advance();
+        }
+        while let Some(iter) = self.iter.get_mut() {
+            if self.inner_first || !iter.is_done() {
+                self.inner_first = false;
```

Comment:
Why do we need `inner_first`? I think we can unconditionally `advance` the inner iterator here, just like `FlatMap` does. That can always be called at least once, and if that sees `is_done()`, then we're moving on to the next `self.iter.advance()` anyway.

Reply:
You are right. I somehow missed the fact that after we advance the outer iterator (`self.iter`) we have not yet called advance on the new inner one and are therefor able to call `advance` on it unconditionally.